src/invocation.rs

```diff
@@ -0,0 +1,117 @@
+use crate::{
+    ability::traits::{Buildable, Command},
+    delegation,
+    delegation::{condition::Condition, Delegate, Delegation},
+    receipt::Receipt,
+    signature,
+    time::Timestamp,
+};
+use did_url::DID;
+use libipld_core::{cid::Cid, ipld::Ipld, link::Link};
+use std::{collections::BTreeMap, fmt::Debug};
+
+pub type Invocation<B> = signature::Envelope<Payload<B>>;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Payload<B: Buildable + Debug> {
+    pub issuer: DID,
+    pub subject: DID,
+    pub audience: Option<DID>,
+
+    pub ability: B,
```

Comment:
So without a task/instruction (not being steadfast about layers here per se, but Task in 0.2 spec), this structure, as is shown and implemented, makes it difficult to have an index to match on a reusable key for replayability. Going off previous versions, this reusable key was composed of:

- `rsc`
- `input`/args/etc
- `op`/(old) ability string, i.e. wasm/run, crud/etc
- `nonce` (nnc)

The nonce as part of this was handy b/c it could separate one instruction/task from the other if you so chose to. With 1.0, `rsc` is now inside ability/run/arguments, which is fine. But, calculating this generic task "index" is key, and nonce should be a part of it?

As per the 1.0 spec, A Task MUST be uniquely defined by the following fields:

[Subject](https://github.com/ucan-wg/invocation/blob/high-level/README.md#subject)
[Command](https://github.com/ucan-wg/invocation/blob/high-level/README.md#command)
[Arguments](https://github.com/ucan-wg/invocation/blob/high-level/README.md#arguments)
[Nonce](https://github.com/ucan-wg/invocation/blob/high-level/README.md#nonce)

For reusability, the `sub` subject may not be handy here, but the rest are. If we don't want this represented within the structure, then it needs to be something that can be rendered from it (function with a known name quantity that we use).

Reply:
> makes it difficult to have an index to match on a reusable key for replayability.

Yeah, we're going to have to calculate our own. This is better for a few reasons, but the two bigs ones are:

* We can pick the serialisation
* Getting people to respect the nonce parameter for idempotent actions was going to be _hard_

Previously, if two people submitted the same invocation, but one was DAG-CBOR and the other was DAG-JSON, we wouldn't have the same index

Reply:
> For reusability, the sub subject may not be handy here, but the rest are

It depends. It doesn't matter for pure Wasm, but it does as soon as effects show up.

Reply:
> As per the 1.0 spec, A Task MUST be uniquely defined by the following fields:

Sorry, that's probably my bad. _All_ of (100%) of the indexing stuff should be removed from the invocation spec. It's an IPVM concern, not a UCAN one.

Reply:
> > For reusability, the sub subject may not be handy here, but the rest are
> 
> It depends. It doesn't matter for pure Wasm, but it does as soon as effects show up (i.e. calling the same function on two agents can yield different results)

Right. I guess, it's not something you always want in the calculation is what I want to say.

Reply:
> > makes it difficult to have an index to match on a reusable key for replayability.
> 
> Yeah, we're going to have to calculate our own. This is better for a few reasons, but the two bigs ones are:
> 
> * We can pick the serialisation
> * Getting people to respect the nonce parameter for idempotent actions was going to be _hard_
> 
> Previously, if two people submitted the same invocation, but one was DAG-CBOR and the other was DAG-JSON, we wouldn't have the same index. We can also determine if we want to include the nonce or not based on metadata about the task being run.

yeah, I think the key is dealing with duplicate tasks once again and that nonce has meaning there if you want to run the same exact thing (for whichever reason).

Reply:
I'm trying to remember back to when this was written, but we probably usually (always?) do want the subject in there, because ultimately they're the one that set the semantics

Reply:
Yeah. It wasn't part of the original task/instruction in 0.1/0.2. It was there, but not included. If it's the case that sub (in like the wasm case) will typically be the same across runs (and machines), then that's fine. Mainly, what works for reusability when we want to allow that.